libs/libskynet/src/registry.ts

```diff
@@ -0,0 +1,195 @@
+// different.
+//
+// Note that changing the keypair tag will also change the resulting datakey.
+// The purpose of the keypair tag is to obfuscate the fact that two registry
+// entries are owned by the same identity. This obfuscation would break if two
+// different public keys were using the same datakey. Changing the datakey does
+// not change the public key.
+function taggedRegistryEntryKeys(
+  seed: Uint8Array,
+  keypairTagStr: string,
+  datakeyTagStr?: string
+): [Ed25519Keypair, Uint8Array, string | null] {
+  if (seed.length !== SEED_BYTES) {
+    return [{} as Ed25519Keypair, new Uint8Array(0), "seed has the wrong length"];
+  }
+  if (keypairTagStr.length > 255) {
+    return [{} as Ed25519Keypair, new Uint8Array(0), "keypairTag must be less than 256 characters"];
+  }
+  // If no datakey tag was provided, use the empty string.
+  if (datakeyTagStr === undefined) {
+    datakeyTagStr = "";
+  }
+
+  // Generate a unique set of entropy using the seed and keypairTag.
+  const keypairTag = new TextEncoder().encode(keypairTagStr);
+  const entropyInput = new Uint8Array(keypairTag.length + seed.length);
+  entropyInput.set(seed, 0);
+  entropyInput.set(keypairTag, seed.length);
+  const keypairEntropy = sha512(entropyInput);
+
+  // Use the seed to dervie the datakey for the registry entry. We use
+  // a different tag to ensure that the datakey is independently random, such
+  // that the registry entry looks like it could be any other registry entry.
+  //
+  // We don't want it to be possible for two different combinations of
+  // tags to end up with the same datakey. If you don't use a length
+  // prefix, for example the tags ["123", "456"] and ["12", "3456"] would
+  // have the same datakey. You have to add the length prefix to the
+  // first tag otherwise you can get pairs like ["6", "4321"] and ["65",
+  // "321"] which could end up with the same datakey.
+  const datakeyTag = new TextEncoder().encode(datakeyTagStr);
+  const datakeyInput = new Uint8Array(seed.length + 1 + keypairTag.length + datakeyTag.length);
+  const keypairLen = new Uint8Array(1);
+  keypairLen[0] = keypairTag.length;
+  datakeyInput.set(seed);
+  datakeyInput.set(keypairLen, seed.length);
+  datakeyInput.set(keypairTag, seed.length + 1);
+  datakeyInput.set(datakeyTag, seed.length + 1 + keypairTag.length);
+  const datakeyEntropy = sha512(datakeyInput);
+
+  // Create the private key for the registry entry.
+  const [keypair, errKPFE] = ed25519KeypairFromEntropy(keypairEntropy.slice(0, 32));
+  if (errKPFE !== null) {
+    return [{} as Ed25519Keypair, new Uint8Array(0), addContextToErr(errKPFE, "unable to derive keypair")];
+  }
+  const datakey = datakeyEntropy.slice(0, 32);
+  return [keypair, datakey, null];
+}
+
+// verifyRegistrySignature will verify the signature of a registry entry.
+function verifyRegistrySignature(
+  pubkey: Uint8Array,
+  datakey: Uint8Array,
+  data: Uint8Array,
+  revision: bigint,
+  sig: Uint8Array
+): boolean {
```

Comment:
this is different because the error message typically isn't useful. In other places where we return an `error` instead of a boolean it's typically because you want to relay some message back to the user. For example, if we're validating a seed, presumably that seed came from a user, and the user is interested in seeing **why** the seed is incorrect if there's an issue.

But when you are validating a signature, the only possibilities are:

1. signing code was wrong (and the sig was given to you by someone else so the error doesn't matter much)
2. malice is involed, which means again the error is not very useful you just want it to fail